### PR TITLE
command/jsonstate: fix inconsistency with resource address

### LINKED ISSUE #24110: Resource addresses in prior state of a json plan lacking indexes
### Terraform Version

```
0.12.20
```

### Terraform Configuration Files
```hcl
resource "null_resource" "cluster" {
  count = 3
}
```

### Expected Behavior
I expect there to be symmetry between prior state and proposed state. The proposed state looks like

```json
{
  "format_version": "0.1",
  "terraform_version": "0.12.20",
  "planned_values": {
    "root_module": {
      "resources": [
        {
          "address": "null_resource.cluster[0]",
          "mode": "managed",
          "type": "null_resource",
          "name": "cluster",
          "index": 0,
          "provider_name": "null",
          "schema_version": 0,
          "values": {
            "triggers": null
          }
        },
...
```

### Actual Behavior
The prior state, if I delete the resources, looks like

```json
  "prior_state": {
    "format_version": "0.1",
    "terraform_version": "0.12.20",
    "values": {
      "root_module": {
        "resources": [
          {
            "address": "null_resource.cluster",
            "mode": "managed",
            "type": "null_resource",
            "name": "cluster",
            "index": 0,
            "provider_name": "null",
            "schema_version": 0,
            "values": {
              "id": "7039102915404397285",
              "triggers": null
            }
          },
...
```

### Steps to Reproduce
1. Use example input
1. `terraform init`
1. `terraform apply`
1. delete resource
1. `tf plan -out terraform.tfplan; tf show -json terraform.tfplan`

Note that the resource addresses in the prior state do not contain an index.

### PR DESCRIPTION
**Pull request by mildwonkey**
Resource addresses in state output were not including index for
instances created with for_each or count, while the index was appearing
in the plan output. This PR fixes that inconsistency, adds tests, and
updates the existing tests.

Fixes #24110

### CODE REVIEW

**Comment by mildwonkey** [comment]
@vancluever that makes sense! The end to end tests, which include config and expected json output, are in the command package: https://github.com/hashicorp/terraform/tree/master/command/testdata/show-json

Does that cover your request? I can (and will) go add a few more complicated examples that include resources created with for_each and count - and let me know if there's anything else you think is missing from those examples.

**Comment by vancluever** [comment]
Yeah e2e would be the prefect place. Did these not need modifying in their current form?

**Comment by mildwonkey** [comment]
The e2e tests did not need updates because none of them used (non-empty) prior state, and the state output is where the non-instanced addressing appeared. (I am adding to that now)

**Comment by mwhooker** [comment]
exciting! Unfortunately I'm unable to test beyond the example in my report, because I just have a json file that reproduces the more complex issue.

File: command/testdata/show-json/multi-resource-update/output.json [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            "outputs": {
+                "test": {
+                    "sensitive": false,
+                    "value": "bar"
+                }
+            },
+            "root_module": {
+                "resources": [
+                    {
+                        "address": "test_instance.test[0]",
```
Comment (mildwonkey) [inline command/testdata/show-json/multi-resource-update/output.json]: @mwhooker I believe this is the important test for you :) And thank you again for reporting the bug!

Comment (mwhooker) [reply command/testdata/show-json/multi-resource-update/output.json]: this looks good! It looks like the code covers this case, but there's also the issue of modules getting lost. I wonder if that's worth capturing in a test? up to you

Comment (mildwonkey) [reply command/testdata/show-json/multi-resource-update/output.json]: > the issue of modules getting lost
Sorry, that's not ringing any bells - is there a related GH issue? 

There are e2e tests that cover modules in configuration, but not prior state. I'll add (at least) one, and if you can share the module issue I can tailor a test to that specifically.

**Pull request merged**